README
======
HTTP 1
======

HTTP1 is a wrapper around httplib to perform HTTP requests in a single call.
For instance, to get PyPI index of packages, you might write::

    import http1
    
    print http1.request('http://pypi.python.org/simple/').body


request() method
================

This method performs an HTTP request. The signature of the request method is
the following::

    request(url, params={}, method='GET', body=None, headers={},
            content_type=None, content_length=True, username=None,
            password=None, capitalize_headers=True)

The parameters are the following:

- url: the URL call, including protocol and parameters (such as
  'http://www.google.com?foo=1&bar=2').
- params: URL parameters as a map, so that {'foo': 1, 'bar': 2} will result
  in an URL ending with '?foo=1&bar=2'.
- method: the HTTP method (such as 'GET' or 'POST'). Defaults to 'GET'.
- body: the body of the request as a string. Defaults to None.
- headers: request headers as a dictionnary. Defaults to '{}'.
- content_type: the content type header of the request. Defauls to None.
- content_length: tells if we should add content length headers to the
  request. Defaults to true.
- username: username while performing basic authentication, must be set
  with password.
- password: password while performing basic authentication, must be set
  with username.
- capitalize_headers: tells if headers should be capitalized (so that their
  names are all like 'Content-Type' for instance).

This method returns the response as a Response object described hereafter.

*NOTE*: to call HTTPS URLs, Python must have been built with SSL support.

Response object
===============

This object encapsulates status code (200, 404, as an integer), message (such
as 'OK', 'Not Found', as a string), headers (as a dictionnary), and body (as a
string).

Releases
========

- 0.1.0 (2012-03-29): First public release.
- 0.1.1 (2012-03-30): Documentation fixes.
- 0.1.2 (2012-03-30): More documentation fixes.
- 0.1.3 (2013-07-05): Fixed bug regarding matrix params.



CHANGELOG.md
Change Log
==========

0.3.2: 2015-04-26
-----------------

Fixed relative redirects (thanks to John Vandenberg).

0.3.1: 2014-05-05
-----------------

Fixed HTTP methods access.

0.3.0: 2014-04-05
-----------------

Added functions for HTTP methods.

0.2.1: 2014-04-05
-----------------

Added TooManyRedirectsException.

0.2.0: 2014-04-05
-----------------

Added option follow_redirect.

0.1.4: 2014-04-04
-----------------

Project structure refactoring, included license in the archive.

0.1.3: 2013-07-05
-----------------

Fixed bug regarding matrix params.

0.1.2: 2012-03-30
-----------------

More documentation fixes.

0.1.1: 2012-03-30
-----------------

Documentation fixes.

0.1.0: 2012-03-29
-----------------

First public release.


test/test_http1.py
#!/usr/bin/env python
# encoding=UTF-8

import os
import sys
sys.path.append(os.path.join(os.path.dirname(__file__), '..', 'src'))
import http1
import unittest

class Test(unittest.TestCase):

    def test_get_ok(self):
        response = http1.request('http://sweetohm.net')
        self.assertEquals(200, response.status)
        self.assertEquals('OK', response.message)
        self.assertTrue('Sweetohm' in response.body)

    def test_status_ko(self):
        response = http1.request('http://sweetohm.net/toto')
        self.assertEquals(404, response.status)
        self.assertEquals('Not Found', response.message)

    def test_get_https_ok(self):
        try:
            headers = {'Accept-Language': 'fr'}
            response = http1.request('https://www.google.fr',
                                     headers=headers)
            self.assertEquals(200, response.status)
            self.assertEquals('OK', response.message)
            self.assertTrue('<title>Google</title>' in response.body)
        except AttributeError:
            # if Python was not built with SSL support
            pass

    def test_redirect(self):
        response = http1.request('http://sweetohm.net/blog/index.html')
        expected = 200
        actual = response.status
        self.assertEqual(expected, actual)
        self.assertTrue('Sweetohm' in response.body)
        response = http1.request('http://sweetohm.net/arc/python-dbapi.pdf',
                                 follow_redirect=False)
        expected = 301
        actual = response.status
        self.assertEqual(expected, actual)
        try:
            http1.request('http://cafebabe.free.fr',
                          max_redirect=0)
        except http1.TooManyRedirectsException:
            pass

    def test_methods(self):
        response = http1.get('http://www.sweetohm.net')
        self.assertEqual(200,response.status)
        self.assertTrue('Sweetohm' in response.body)
        response = http1.head('http://sweetohm.net/arc/python-dbapi.pdf',
                              follow_redirect=False)
        self.assertEqual(301, response.status)
        self.assertEqual('http://sweetohm.net/pdf/python-dbapi.pdf',
                         response.headers['Location'])

    def test_relative_redirect(self):
        response = http1.get('http://httpbin.org/redirect/2')
        self.assertEqual(200, response.status)


# Run unit tests when started on command line
if __name__ == '__main__':
    unittest.main()



http1/test_http1.py
#!/usr/bin/env python
# encoding=UTF-8

#pylint: disable=W0403
import http1
import unittest

class Test(unittest.TestCase):

    def test_get_ok(self):
        response = http1.request('http://sweetohm.net')
        self.assertEquals(200, response.status)
        self.assertEquals('OK', response.message)
        self.assertTrue('Sweetohm' in response.body)

    def test_status_ko(self):
        response = http1.request('http://sweetohm.net/toto')
        self.assertEquals(404, response.status)
        self.assertEquals('Not Found', response.message)

    def test_get_https_ok(self):
        try:
            headers = {'Accept-Language': 'fr'}
            response = http1.request('https://www.google.fr',
                                     headers=headers)
            self.assertEquals(200, response.status)
            self.assertEquals('OK', response.message)
            self.assertTrue('<title>Google</title>' in response.body)
        except AttributeError:
            # if Python was not built with SSL support
            pass

    def test_redirect(self):
        response = http1.request('http://sweetohm.net/blog/index.html')
        expected = 200
        actual = response.status
        self.assertEqual(expected, actual)
        self.assertTrue('Sweetohm' in response.body)
        response = http1.request('http://sweetohm.net/arc/python-dbapi.pdf',
                                 follow_redirect=False)
        expected = 301
        actual = response.status
        self.assertEqual(expected, actual)
        try:
            http1.request('http://cafebabe.free.fr',
                          max_redirect=0)
        except http1.TooManyRedirectsException:
            pass

    def test_methods(self):
        response = http1.get('http://www.sweetohm.net')
        self.assertEqual(200,response.status)
        self.assertTrue('Sweetohm' in response.body)
        response = http1.head('http://sweetohm.net/arc/python-dbapi.pdf',
                              follow_redirect=False)
        self.assertEqual(301, response.status)
        self.assertEqual('http://sweetohm.net/pdf/python-dbapi.pdf',
                         response.headers['Location'])

    def test_relative_redirect(self):
        response = http1.get('http://httpbin.org/redirect/2')
        self.assertEqual(200, response.status)


# Run unit tests when started on command line
if __name__ == '__main__':
    unittest.main()



http1/http1.py
# encoding: UTF-8

"""
Module to perform HTTP requests. To do so:

  import http

  response = http.request('www.google.com')
  print 'Status: %s (%s)' % (response.status, response.message)
  print 'Headers: %s' % response.headers
  print 'Body: %s' % response.body.strip()

"""

import urllib
import httplib
import urlparse
import base64

class Response(object):
    """HTTP response to encapsulates status code (200, 404, as an integer),
    message (such as 'OK', 'Not Found', as a string), headers (as a
    dictionnary), and body (as a string)."""

    def __init__(self, status, message, headers={}, body=None):
        self.status = status
        self.message = message
        self.headers = headers
        self.body = body

    def __str__(self):
        if self.body:
            _body = str(self.body).strip().replace('\n', '\\n')
            if len(_body) > 100:
                _body = _body[:97]+'...'
        else:
            _body = ''
        return "Response(status=%s, message='%s', headers=%s, body='%s')" %\
            (self.status, self.message, self.headers, _body)


class TooManyRedirectsException(Exception):

    pass


def request(url, params={}, method='GET', body=None, headers={},
            content_type=None, content_length=True,
            username=None, password=None, capitalize_headers=True,
            follow_redirect=True, max_redirect=3):
    """Perform a http_request:
    - url: the URL call, including protocol and parameters (such as
      'http://www.google.com?foo=1&bar=2').
    - params: URL parameters as a map, so that {'foo': 1, 'bar': 2} will result
      in an URL ending with '?foo=1&bar=2'.
    - method: the HTTP method (such as 'GET' or 'POST'). Defaults to 'GET'.
    - body: the body of the request as a string. Defaults to None.
    - headers: request headers as a dictionnary. Defaults to '{}'.
    - content_type: the content type header of the request. Defauls to None.
    - content_length: tells if we should add content length headers to the
      request. Defaults to true.
    - username: username while performing basic authentication, must be set
      with password.
    - password: password while performing basic authentication, must be set
      with username.
    - capitalize_headers: tells if headers should be capitalized (so that their
      names are all like 'Content-Type' for instance).
    - follow_redirect: tells if http1 should follow redirections (status codes
      3xx). Defaults to True.
    - max_redirect: maximum number of redirections to follow. If there are too
      many redirects, a TooManyRedirectsException is raised. Defaults to 3.
    Returns the response as a Response object.
    Raise TooManyRedirectsException.
    NOTE: to call HTTPS URLs, Python must have been built with SSL support."""
    _urlparts = urlparse.urlparse(url)
    _host = _urlparts.netloc
    _matrix_params = _urlparts.params
    _params = ''
    if len(_urlparts.query) > 0:
        _params = _urlparts.query
    if len(params) > 0:
        if len(_params) > 0:
            _params += '&'
        _params += urllib.urlencode(params)
    _path = _urlparts.path
    if _matrix_params:
        _path += ';%s' % _matrix_params
    if len(_params) > 0:
        _path += '?'
        _path += _params
    _https = (_urlparts.scheme == 'https')
    _headers = {}
    for _name in headers:
        _headers[str(_name)] = str(headers[_name])
    if content_type:
        _headers['Content-Type'] = str(content_type)
    if content_length:
        if body:
            _headers['Content-Length'] = str(len(body))
        else:
            _headers['Content-Length'] = '0'
    if username and password:
        authorization = "Basic %s" % base64.b64encode(("%s:%s" % (username, password)))
        _headers['Authorization'] = authorization
    _capitalized_headers = {}
    if capitalize_headers:
        for _name in _headers:
            _capitalized = '-'.join([s.capitalize() for s in _name.split('-')])
            _capitalized_headers[_capitalized] =_headers[_name]
        _headers = _capitalized_headers
    if _https:
        connection = httplib.HTTPSConnection(_host)
    else:
        connection = httplib.HTTPConnection(_host)
    connection.request(method, _path, body, _headers)
    _response = connection.getresponse()
    # method getheaders() not available in Python 2.2.1
    _response_headers = {}
    _pairs = _response.msg.items()
    if _pairs:
        for _pair in _pairs:
            _name = _pair[0]
            _value = _pair[1]
            if capitalize_headers:
                _name = '-'.join([s.capitalize() for s in _name.split('-')])
            _response_headers[_name] = _value
    if _response.status >= 300 and _response.status < 400 and \
        follow_redirect:
        if max_redirect <= 0:
            raise TooManyRedirectsException
        location = urlparse.urljoin(url, _response_headers['Location'])
        return request(url=location, params=params, method=method,
                       body=body, headers=headers,
                       content_type=content_type,
                       content_length=content_length,
                       username=username, password=password,
                       capitalize_headers=capitalize_headers,
                       follow_redirect=True,
                       max_redirect=max_redirect-1)
    return Response(status=_response.status,
                    message=_response.reason,
                    headers=_response_headers,
                    body=_response.read())


def get(*args, **kwargs):
    return request(*args, method='GET', **kwargs)


def head(*args, **kwargs):
    return request(*args, method='HEAD', **kwargs)


def post(*args, **kwargs):
    return request(*args, method='POST', **kwargs)


def put(*args, **kwargs):
    return request(*args, method='PUT', **kwargs)


def delete(*args, **kwargs):
    return request(*args, method='DELETE', **kwargs)


def connect(*args, **kwargs):
    return request(*args, method='CONNECT', **kwargs)


def options(*args, **kwargs):
    return request(*args, method='OPTIONS', **kwargs)


def trace(*args, **kwargs):
    return request(*args, method='TRACE', **kwargs)
